Assets/MixedRealityToolkit.Services/InputSimulation/SimulatedGestureHand.cs

```diff
@@ -143,126 +144,178 @@ protected override void UpdateInteractions(SimulatedHandData handData)
 
                                 SelectDownStartTime = Time.time;
                                 cumulativeDelta = Vector3.zero;
+
+                                TryStartManipulation();
                             }
                             else
                             {
                                 InputSystem?.RaiseOnInputUp(InputSource, ControllerHandedness, Interactions[i].MixedRealityInputAction);
 
-                                // Stop gestures
-                                CompleteHoldGesture();
-                                CompleteManipulationNavigationGesture();
+                                // Stop active gestures
+                                TryCompleteHold();
+                                TryCompleteManipulation();
+                                TryCompleteNavigation();
                             }
                         }
                         else if (Interactions[i].BoolData)
                         {
-                            if (!manipulationInProgress)
+                            if (manipulationInProgress)
                             {
-                                if (cumulativeDelta.magnitude > manipulationStartThreshold)
-                                {
-                                    CancelHoldGesture();
-                                    StartManipulationNavigationGesture();
-                                }
-                                else if (!holdInProgress)
-                                {
-                                    float time = Time.time;
-                                    if (time >= SelectDownStartTime + holdStartDuration)
-                                    {
-                                        StartHoldGesture();
-                                    }
-                                }
+                                UpdateManipulation();
                             }
-                            else
+                            if (navigationInProgress)
+                            {
+                                UpdateNavigation();
+                            }
+
+                            if (cumulativeDelta.magnitude > navigationStartThreshold)
```

Comment:
So here's a really dumb question from me, but looking at how cumulativeDelta is updated:

cumulativeDelta += currentPosition - lastPosition;

It seems like it would be possible to actually move your hand back into a position such that the cumulative delta ends up being 0,0,0 again, which would in theory start the hold here.

Is this not the case (i.e. is there some abs() going on elsewhere?)

Reply:
Yes, this can happen in theory (have never seen it happen in practice though, not even with the simulated hands). Note that i've copied this from the HTK implementation.

The cumulativeDelta is also used for determining the navigation rails when they are enabled, which requires keeing a vector. The function doesn't look very robust though, we may want to revisit this at some point - if anyone actually uses it.